### ISSUE TITLE
Email pattern is not compliant

### ISSUE THREAD

**Issue by mjeanroy** [keycloak/keycloak#20316]
### Before reporting an issue

- [X] I have searched existing issues
- [X] I have reproduced the issue with the latest release

### Area

account/api

### Describe the bug

The email validator is not compatible with the latest update of RFC 5322: [RFC 6532](https://www.rfc-editor.org/rfc/rfc6532#section-3.2).

More specifically, email address containing accentuated characters are considered invalid by Keycloak, which should be valid nowadays.

A more compliant regex:

```
private static final String EMAIL_PATTERN = "[^<>()[\\\\]\\\\.,;:\\s@\"]+(\\.[^<>()[\\\\]\\\\.,;:\\s@\"]+)*@([a-zA-Z0-9-]+\\.[a-zA-Z0-9-.]+)";
```

Note: This one is not compliant with IP address used as domain name.

I can open a PR to replace the current `EMAIL_PATTERN` if you think it's worth it.

### Version

21.0.1

### Expected behavior

Creating a user with an email containing an accentuated character should works.

### Actual behavior

Creating a user with an email containing an accentuated character fail with an HTTP 400 error:

```
{"errorMessage":"Invalid email address."}
```

### How to Reproduce?

Create a user with an email containing an accentuated character.

**Comment by mathieutu** [comment]
💯
FYI, Gmail [supports](https://gmail.googleblog.com/2014/08/a-first-step-toward-more-global-email.html) accented and non Latin characters since 2014.

**Comment by sschu** [comment]
Is that new pattern compatible with the old one or are there mail addresses considered valid with the old one but invalid with the new one?

**Comment by mjeanroy** [comment]
Thanks for your response @sschu!

> Is that new pattern compatible with the old one

In theory, yes, but if you think it's worth opening a PR, I'll add a lot of tests to make sure it works fine (and that it does not break existing tests).

Note that this pattern is not 100% compliant with the RFC as you can (in theory) use IP instead of a domain name (`like john.doe@x.x.x.x`). Supporting this case will add a lot of complexity, except if we choose a different solution than using a pattern (like [this one](https://github.com/apache/commons-validator/blob/35f43a63bd1689961ada7fcea4d3cab81747b5d7/src/main/java/org/apache/commons/validator/routines/EmailValidator.java#L171)).

What could be great is to make sure this parameter is configurable, maybe using a configuration option, so it becomes an explicit opt-in option (and thus, do not change anything by default).

Another option would be adding a custom spi overriding the default email validator. I tried this approach, but the email validator is hardcoded and cannot be bypassed. Taking a step back, that would be my favorite approach:
- Almost no impact on the current validation flow.
- Keycloak users are free to use their own validators for emails.

**Comment by mposolda** [comment]
@mjeanroy @sschu Thanks for the report and discussion. The PR would be welcome.

IMO for now it is sufficient to make sure that `Validation.isEmailValid(String email)` method supports the same email addresses as before (so there is no regression with previously supported email addresses being now considered invalid). So ideally new address formats can be included in addition to that.

If it's possible to do this just by modifying EMAIL_PATTERN regex, it would be ideal.

+1 that backend and UI validation should be aligned.

AFAIK with the user-profile feature enabled, there is a plan to have configurable support for email address validation. Without user-profile feature, it is not configurable, but that's ok for now. In other words, this PR should not introduce any "configurable validation" for the case without user-profile as the plan is to rely on user-profile in the future and make it supported feature.

Note: There is a test `ValidationTest.isEmailValid` for this, so hopefully you can just re-use that one and add new things to it.

**Comment by jonkoops** [comment]
Just FYI, on the front-end we are most likely to use `<input type="email">`, which comes with it's own set of [validation](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/email#validation). Perhaps the regular expression there can be of some help?

**Comment by ivanjecim** [comment]
We are also using accentuated characters in email addresses so we cannot upgrade to Keycloak 21 until email regex is updated.

I saw in tests that current EMAIL_PATTERN considers `1@A` and `A@something` as valid email addresses, but I think it's not correct.

So I suggest 3 options:

1. add only support for accentuated characters before @:
	`private static final Pattern EMAIL_PATTERN = Pattern.compile("[^<>()\\[\\]\\\\.,;:\\s@\"]+(\\.[^<>()\\[\\]\\\\.,;:\\s@\"]+)*@[a-zA-Z0-9-]+(\\.[a-zA-Z0-9-]+)*");`

2. fix also domain part of email address:

	`private static final Pattern EMAIL_PATTERN = Pattern.compile("[^<>()\\[\\]\\\\.,;:\\s@\"]+(\\.[^<>()\\[\\]\\\\.,;:\\s@\"]+)*@([a-zA-Z0-9-]+\\.)+[a-zA-Z]{2,}");`

3. allow also IP address after @, but UI does not allow it:

	`private static final Pattern EMAIL_PATTERN = Pattern.compile("[^<>()\\[\\]\\\\.,;:\\s@\"]+(\\.[^<>()\\[\\]\\\\.,;:\\s@\"]+)*@([0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3})|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,})");`
	
	
IMO option 2 is the nearest to the validation in the UI. Option 1 should not affect existing emails but also allows incorrect ones as mentioned above.

**Comment by jonkoops** [comment]
What about using [Jakarta Mail](https://github.com/jakartaee/mail-api)?

**Comment by mihaiborbea** [comment]
I also discovered that i am able to create users (via API) with invalid emails such as "email@withoutanydot". Is this expected behavior? I would expect not.

**Comment by mathieutu** [comment]
> email@withoutanydot
This is actually a valid email 🙂

**Comment by mihaiborbea** [comment]
> > email@withoutanydot
> > This is actually a valid email 🙂

Well yes it somehow is, but i was also looking into the RFC 5322 which states that domain part should contain at least a dot.

**Comment by rmartinc** [comment]
I think this was finally fixed in https://github.com/keycloak/keycloak/issues/20878. So at least at server side the fix is OK in current 22.0.x versions. Closing as duplicated.
